### PR TITLE
Support required text and optional text configuration.

### LINKED ISSUE #1954: Show required text if the question must be answered. Otherwise, provide configuration to support optional text.
**Is your feature request related to a problem? Please describe.**
As per the discussion with @shelaghm and @jingtang10, and considering the input from the external community, it has been decided that
1. If a question item is required, then the required text will be shown in the form just below the view. This will be the default behavior.
2. Also support the configuration to show optional text for optional questions instead of the default behavior.

#544 

**Describe the solution you'd like**

**Describe alternatives you've considered**

**Additional context**

**Would you like to work on the issue?**
Please state if this issue should be assigned to you or who you think could help to solve this issue.

### PR DESCRIPTION
**Pull request by santosh-pingle**
**IMPORTANT: All PRs must be linked to an issue (except for extremely trivial and straightforward changes).**

Fixes #1954 

**Description**

-  Show required text to the beginning of the instructions if views are non textInputLayout views.
-  Show required text as helper text if view are InputTextLayout views.
-  Support optional text configuration
-  If optional text is configured, then show optional text at the end of the question text.
-  If question text is not present then show optional text as helper text.

**Alternative(s) considered**
Have you considered any alternatives? And if so, why have you chosen the approach in this PR?

**Type**
Choose one: Feature 

**Screenshots (if applicable)**

**Checklist**
- [x] I have read and acknowledged the [Code of conduct](https://github.com/google/android-fhir/blob/master/docs/CODE_OF_CONDUCT.md).
- [x] I have read the [Contributing](https://github.com/google/android-fhir/wiki/Contributing) page.
- [x] I have signed the Google [Individual CLA](https://cla.developers.google.com/about/google-individual), or I am covered by my company's [Corporate CLA](https://cla.developers.google.com/about/google-corporate ).
- [x] I have discussed my proposed solution with code owners in the linked issue(s) and we have agreed upon the general approach.
- [x] I have run `./gradlew spotlessApply` and `./gradlew spotlessCheck` to check my code follows the style guide of this project.
- [x] I have run `./gradlew check` and `./gradlew connectedCheck` to test my changes locally.
- [x] I have built and run the demo app(s) to verify my change fixes the issue and/or does not break the demo app(s).

### CODE REVIEW

**Comment by santosh-pingle** [comment]
@shelaghm 
Required text : 
Screenshot 2023-04-13 at 12 30 06 PM

Optional text : 
Screenshot 2023-04-13 at 12 24 13 PM

**Comment by shelaghm** [comment]
This looks good to me. Thanks @santosh-pingle

**Comment by shelaghm** [comment]
1. To be consistent between the two, I recommend we use the instruction text field to append required/optional. In paginated layout the (required) or (optional) text is very large and I think it's not THAT important to display. In default layout it works better than in paginated as the text proportions are more subtle.
2. Breaking it into two lines is a nice suggestion. @santosh-pingle Do you think we could do a line-break?

I've gone through and mocked up all the options, so we can be sure we are talking about the same options. I'll also post link in group chat as it's a lot! 
Screenshot 2023-04-24 at 16 40 13

**Comment by shelaghm** [comment]
Feedback from @mberg 
To confirm does this mean the sdc will no longer support * or will that be supported as an option? I would want to keep the option to not have to display optional on every question. I feel like it’s a bit overwhelming to see it on every question.  Odk and most of the form libraries just use the *.  I think a lot of people are used to the pattern but I see why required text is a nice option too. I would want to disable optional on all questions though as I feel that becomes a bit overwhelming.

**Comment by santosh-pingle** [comment]
@shelaghm 
Thanks for sharing the details.
The only pending questions are: what will be the default behavior, and what will be supported from the given table?

**Comment by shelaghm** [comment]
@santosh-pingle I recommend that the default is `2 - Required` and that is appended to `instructions or help text`. So that is 2A, 2B, 2C. Default should be separate lines for Required/Optional and additional instructions

I suggest we make the following configurable:
- Optional (3)
- Asterisk (1)

I'm not sure if we need to add ability to also append the required/optional text behind the question title instead of in the instructions/help field. 

@jingtang10 what do you think? Let's discuss

File: datacapture/src/main/java/com/google/android/fhir/datacapture/QuestionnaireViewModel.kt [hunk 1]
```diff
@@ -218,6 +218,10 @@ internal class QuestionnaireViewModel(application: Application, state: SavedStat
   /** Flag to show/hide submit button. Default is true. */
   private var shouldShowSubmitButton = state[QuestionnaireFragment.EXTRA_SHOW_SUBMIT_BUTTON] ?: true
 
+  /** Flag to control if the optional text is being added to the end of the question text. */
+  private var markOptionalQuestionText =
+    state[QuestionnaireFragment.EXTRA_MARK_OPTIONAL_QUESTION] ?: false
```
Comment (dubdabasoduba) [inline datacapture/src/main/java/com/google/android/fhir/datacapture/QuestionnaireViewModel.kt:223]: @santosh-pingle While testing I changed this to `true` and then re-run the code. The optional text was shown but even on the required fields. I haven't investigated much but I think we only need to show the optional text on the non-required fields. 

Feel free to update me incase I missed something.

**Review by dubdabasoduba — CHANGES_REQUESTED** [review]

File: datacapture/src/main/java/com/google/android/fhir/datacapture/QuestionnaireViewModel.kt (hunk 1 above)
Comment (santosh-pingle) [reply datacapture/src/main/java/com/google/android/fhir/datacapture/QuestionnaireViewModel.kt:223]: I am going to make changes as per above mentioned comment, once it is done I will update you. thanks!

Comment (santosh-pingle) [reply datacapture/src/main/java/com/google/android/fhir/datacapture/QuestionnaireViewModel.kt]: @dubdabasoduba Can you please test the latest changes I have made?

Comment (santosh-pingle) [reply datacapture/src/main/java/com/google/android/fhir/datacapture/QuestionnaireViewModel.kt]: Please update below existing functions to `DemoQuestionnaireFragment` in the catalog app while testing the changes.

```
  private fun addQuestionnaireFragment() {
    viewLifecycleOwner.lifecycleScope.launch {
      if (childFragmentManager.findFragmentByTag(QUESTIONNAIRE_FRAGMENT_TAG) == null) {
        childFragmentManager.commit {
          setReorderingAllowed(true)
          add(
            R.id.container,
            QuestionnaireFragment.builder()
              .setQuestionnaire(viewModel.getQuestionnaireJson())
              .showOptionalText(true)
              .showAsterisk(true)
              .showRequiredText(true)
              .build(),
            QUESTIONNAIRE_FRAGMENT_TAG
          )
        }
      }
    }
  }

private fun replaceQuestionnaireFragmentWithQuestionnaireJson() {
    // TODO: remove check once all files are added
    if (args.questionnaireFileWithValidationPathKey.isNullOrEmpty()) {
      return
    }
    viewLifecycleOwner.lifecycleScope.launch {
      val questionnaireJsonString =
        if (isErrorState) {
          viewModel.getQuestionnaireWithValidationJson()
        } else {
          viewModel.getQuestionnaireJson()
        }
      childFragmentManager.commit {
        setReorderingAllowed(true)
        replace(
          R.id.container,
          QuestionnaireFragment.builder()
            .setQuestionnaire(questionnaireJsonString)
            .showOptionalText(true)
            .showAsterisk(true)
            .showRequiredText(true)
            .build(),
          QUESTIONNAIRE_FRAGMENT_TAG
        )
      }
    }
  }

**Comment by santosh-pingle** [comment]
@shelaghm 

Screenshot 2023-05-05 at 10 56 26 AM
Screenshot 2023-05-05 at 10 56 17 AM
Screenshot 2023-05-05 at 10 56 04 AM
Screenshot 2023-05-05 at 10 55 56 AM
Screenshot 2023-05-05 at 10 55 42 AM
Screenshot 2023-05-05 at 10 55 33 AM
Screenshot 2023-05-05 at 10 55 19 AM
Screenshot 2023-05-05 at 10 55 12 AM
Screenshot 2023-05-05 at 10 54 59 AM

**Comment by dubdabasoduba** [comment]
@santosh-pingle I tested all three options and they seem to work well. I however have a small design request for the Asterisk display

Screenshot 2023-05-08 at 14 08 05

I think we need some spacing between the text and the (*). I think the (*) should have a red text colour. 

cc: @shelaghm

**Comment by dubdabasoduba** [comment]
@shelaghm should these 3 options be mutually exclusive? Right now you can how all can be turned on at the same time. 

Screenshot 2023-05-08 at 16 39 14

**Comment by shelaghm** [comment]
> I think we need some spacing between the text and the Asterisk. 

Yes, agreed. @santosh-pingle we should add one space before the asterisk. 

> I think the Asterisk should have a red text colour.

From our usability testing we learned that the red color on the asterisk was confusing, some people thought it was an error because it was red. So I don't recommend making the asterisk red. 

> should these 3 options be mutually exclusive? Right now you can how all can be turned on at the same time.
Yes, you can turn on all of these at the same time, because you might want to show the required text and the asterisk at the same time. 

@dubdabasoduba @santosh-pingle @jingtang10

File: datacapture/src/main/res/values/strings.xml [hunk 2]
```diff
@@ -116,4 +116,10 @@
     >Number must be between <xliff:g id="min_value">%1$s</xliff:g> and <xliff:g
             id="max_value"
         >%2$s</xliff:g></string>
+    <string name="optional_text">\u0020(optional)</string>
+    <string name="optional_helper_text">Optional</string>
+    <string name="required_in_instructions">Required.\u0020</string>
```
Comment (jingtang10) [inline datacapture/src/main/res/values/strings.xml:121]: remove

File: datacapture/src/main/java/com/google/android/fhir/datacapture/QuestionnaireViewModel.kt [hunk 3]
```diff
@@ -219,6 +219,16 @@ internal class QuestionnaireViewModel(application: Application, state: SavedStat
   /** Flag to show/hide submit button. Default is true. */
   private var shouldShowSubmitButton = state[QuestionnaireFragment.EXTRA_SHOW_SUBMIT_BUTTON] ?: true
 
+  /** Flag to control whether asterisk text is shown. */
+  private var showAsterisk = state[QuestionnaireFragment.EXTRA_SHOW_ASTERISK_TEXT] ?: false
+
+  /** Flag to control whether required text is shown. */
+  private var showRequiredText = state[QuestionnaireFragment.EXTRA_SHOW_REQUIRED_TEXT] ?: true
```
Comment (jingtang10) [inline datacapture/src/main/java/com/google/android/fhir/datacapture/QuestionnaireViewModel.kt:226]:
```suggestion
  private val showRequiredText = state[QuestionnaireFragment.EXTRA_SHOW_REQUIRED_TEXT] ?: true
```

File: datacapture/src/main/java/com/google/android/fhir/datacapture/QuestionnaireViewModel.kt [hunk 4]
```diff
@@ -219,6 +219,16 @@ internal class QuestionnaireViewModel(application: Application, state: SavedStat
...
+
+  /** Flag to control whether optional text is shown. */
+  private var markOptionalQuestionText =
```
Comment (jingtang10) [inline datacapture/src/main/java/com/google/android/fhir/datacapture/QuestionnaireViewModel.kt:229]:
```suggestion
  private val markOptionalQuestionText =
```

Comment (jingtang10) [inline datacapture/src/main/java/com/google/android/fhir/datacapture/QuestionnaireViewModel.kt:229]:
```suggestion
  private var showOptionalText =
```

File: datacapture/src/main/java/com/google/android/fhir/datacapture/views/QuestionnaireViewItem.kt [hunk 5]
```diff
@@ -81,7 +82,10 @@ data class QuestionnaireViewItem(
       emptyList()
     },
   internal val draftAnswer: Any? = null,
-  internal val enabledDisplayItems: List<Questionnaire.QuestionnaireItemComponent> = emptyList()
+  internal val enabledDisplayItems: List<Questionnaire.QuestionnaireItemComponent> = emptyList(),
+  internal val showAsterisk: Boolean = false,
+  internal val showRequiredText: Boolean = false,
+  internal val showOptionalText: Boolean = false,
```
Comment (jingtang10) [inline datacapture/src/main/java/com/google/android/fhir/datacapture/views/QuestionnaireViewItem.kt:88]: wrap in one data class

File: datacapture/src/main/java/com/google/android/fhir/datacapture/views/HeaderView.kt [hunk 6]
```diff
@@ -49,7 +50,7 @@ internal class HeaderView(context: Context, attrs: AttributeSet?) : LinearLayout
       questionnaireItem = questionnaireViewItem.questionnaireItem
     )
     prefix.updateTextAndVisibility(questionnaireViewItem.questionnaireItem.localizedPrefixSpanned)
-    question.updateTextAndVisibility(questionnaireViewItem.questionnaireItem.localizedTextSpanned)
+    appendAsteriskToQuestionText(question, questionnaireViewItem)
```
Comment (jingtang10) [inline datacapture/src/main/java/com/google/android/fhir/datacapture/views/HeaderView.kt:53]: can you keep calling `question.updateTextAndVisibility` here but just call it with a different param?

File: datacapture/src/main/java/com/google/android/fhir/datacapture/extensions/MoreHeaderViews.kt [hunk 7]
```diff
@@ -68,3 +77,74 @@ internal fun initHelpViews(
...
+  questionnaireViewItem.questionnaireItem.localizedTextSpanned?.let {
+    builder.append(questionnaireViewItem.questionnaireItem.localizedTextSpanned)
+  }
+  if (questionnaireViewItem.showAsterisk &&
+      questionnaireViewItem.questionnaireItem.required &&
+      !questionnaireViewItem.questionnaireItem.localizedTextSpanned.isNullOrEmpty()
+  ) {
+    builder.append(questionTextView.context.applicationContext.getString(R.string.space_asterisk))
+  }
+  questionTextView.updateTextAndVisibility(builder)
```
Comment (jingtang10) [inline datacapture/src/main/java/com/google/android/fhir/datacapture/extensions/MoreHeaderViews.kt:135]:
```suggestion
  return SpannableStringBuilder().apply {
    ...
  }
```

File: datacapture/src/main/java/com/google/android/fhir/datacapture/extensions/MoreHeaderViews.kt [hunk 8]
```diff
@@ -68,3 +77,74 @@ internal fun initHelpViews(
...
+  val end = length
+  setSpan(ForegroundColorSpan(color), start, end, Spannable.SPAN_EXCLUSIVE_EXCLUSIVE)
+}
+
+@ColorInt
+private fun Context.getColorFromAttr(@AttrRes attrColor: Int): Int {
+  val typedValue = TypedValue()
+  theme.resolveAttribute(attrColor, typedValue, true)
+  return typedValue.data
+}
```
Comment (jingtang10) [inline datacapture/src/main/java/com/google/android/fhir/datacapture/extensions/MoreHeaderViews.kt:150]: remove

File: datacapture/src/main/res/layout/header_view.xml [hunk 9]
```diff
@@ -58,14 +58,29 @@
...
     </LinearLayout>
+
+    <TextView
+        android:id="@+id/required_optional_text"
+        style="?attr/questionnaireSubtitleTextStyle"
+        android:layout_width="wrap_content"
+        android:layout_height="wrap_content"
+        android:layout_marginTop="@dimen/help_container_margin_top"
+        android:layout_gravity="center_vertical"
+        android:text="@string/required"
```
Comment (jingtang10) [inline datacapture/src/main/res/layout/header_view.xml:69]: remove this as a default?

File: datacapture/src/main/java/com/google/android/fhir/datacapture/views/HeaderView.kt [hunk 10]
```diff
@@ -74,4 +75,26 @@ internal class HeaderView(context: Context, attrs: AttributeSet?) : LinearLayout
       }
     errorTextView.text = errorText
   }
+
+  /** Shows an required text in the header. */
```
Comment (jingtang10) [inline datacapture/src/main/java/com/google/android/fhir/datacapture/views/HeaderView.kt:79]: update the doc

File: datacapture/src/main/java/com/google/android/fhir/datacapture/views/factories/DateTimePickerViewHolderFactory.kt [hunk 11]
```diff
@@ -159,7 +163,17 @@ internal object DateTimePickerViewHolderFactory :
...
+              val validationMessage = validationResult.getSingleStringValidationMessage()
+              if (questionnaireViewItem.questionnaireItem.required &&
+                  questionnaireViewItem.showRequiredText
+              ) {
+                dateInputLayout.context.getString(R.string.required_text_and_new_line) +
+                  validationMessage
+              } else {
+                validationMessage
+              }
+            }
```
Comment (jingtang10) [inline datacapture/src/main/java/com/google/android/fhir/datacapture/views/factories/DateTimePickerViewHolderFactory.kt:176]: extract this and reuse

**Review by jingtang10 — CHANGES_REQUESTED** [review]
thanks @santosh-pingle

**Comment by santosh-pingle** [comment]
Resolving review comments.

File: datacapture/src/main/java/com/google/android/fhir/datacapture/views/HeaderView.kt [hunk 12]
```diff
@@ -74,4 +75,26 @@ internal class HeaderView(context: Context, attrs: AttributeSet?) : LinearLayout
...
+  fun showRequiredOrOptionalTextInHeaderView(questionnaireViewItem: QuestionnaireViewItem) {
```
[the remark this answers is not in the document]
Comment (santosh-pingle) [reply datacapture/src/main/java/com/google/android/fhir/datacapture/views/HeaderView.kt:80]: Required/Optional text is shown as helper text for views like EditTextViewHolderFactory, DatePickerViewHolderFactory etc, therefore it is not called in the bind() of the HeaderView.

File: datacapture/src/main/java/com/google/android/fhir/datacapture/views/factories/DateTimePickerViewHolderFactory.kt (hunk 11 above)
Comment (santosh-pingle) [reply datacapture/src/main/java/com/google/android/fhir/datacapture/views/factories/DateTimePickerViewHolderFactory.kt]: done.

File: datacapture/src/main/java/com/google/android/fhir/datacapture/views/HeaderView.kt (hunk 10 above)
Comment (santosh-pingle) [reply datacapture/src/main/java/com/google/android/fhir/datacapture/views/HeaderView.kt]: done.

File: datacapture/src/main/res/layout/header_view.xml (hunk 9 above)
Comment (santosh-pingle) [reply datacapture/src/main/res/layout/header_view.xml]: done.

File: datacapture/src/main/java/com/google/android/fhir/datacapture/extensions/MoreHeaderViews.kt (hunk 8 above)
Comment (santosh-pingle) [reply datacapture/src/main/java/com/google/android/fhir/datacapture/extensions/MoreHeaderViews.kt]: done.

File: datacapture/src/main/java/com/google/android/fhir/datacapture/extensions/MoreHeaderViews.kt (hunk 7 above)
Comment (santosh-pingle) [reply datacapture/src/main/java/com/google/android/fhir/datacapture/extensions/MoreHeaderViews.kt]: done.

File: datacapture/src/main/java/com/google/android/fhir/datacapture/views/HeaderView.kt (hunk 6 above)
Comment (santosh-pingle) [reply datacapture/src/main/java/com/google/android/fhir/datacapture/views/HeaderView.kt]: done.

File: datacapture/src/main/java/com/google/android/fhir/datacapture/views/QuestionnaireViewItem.kt (hunk 5 above)
Comment (santosh-pingle) [reply datacapture/src/main/java/com/google/android/fhir/datacapture/views/QuestionnaireViewItem.kt]: done.

File: datacapture/src/main/java/com/google/android/fhir/datacapture/QuestionnaireViewModel.kt (hunk 4 above)
Comment (santosh-pingle) [reply datacapture/src/main/java/com/google/android/fhir/datacapture/QuestionnaireViewModel.kt]: done.

Comment (santosh-pingle) [reply datacapture/src/main/java/com/google/android/fhir/datacapture/QuestionnaireViewModel.kt]: done.

File: datacapture/src/main/java/com/google/android/fhir/datacapture/QuestionnaireViewModel.kt (hunk 3 above)
Comment (santosh-pingle) [reply datacapture/src/main/java/com/google/android/fhir/datacapture/QuestionnaireViewModel.kt]: done.

File: datacapture/src/main/java/com/google/android/fhir/datacapture/QuestionnaireViewModel.kt [hunk 13]
```diff
@@ -219,6 +219,16 @@ internal class QuestionnaireViewModel(application: Application, state: SavedStat
   /** Flag to show/hide submit button. Default is true. */
   private var shouldShowSubmitButton = state[QuestionnaireFragment.EXTRA_SHOW_SUBMIT_BUTTON] ?: true
 
+  /** Flag to control whether asterisk text is shown. */
+  private var showAsterisk = state[QuestionnaireFragment.EXTRA_SHOW_ASTERISK_TEXT] ?: false
```
[the remark this answers is not in the document]
Comment (santosh-pingle) [reply datacapture/src/main/java/com/google/android/fhir/datacapture/QuestionnaireViewModel.kt]: done.

File: datacapture/src/main/res/values/strings.xml [hunk 14]
```diff
@@ -116,4 +116,10 @@
...
+    <string name="required">Required</string>
+    <string name="required_text_and_new_line">Required \n</string>
```
[the remark this answers is not in the document]
Comment (santosh-pingle) [reply datacapture/src/main/res/values/strings.xml]: done.

File: datacapture/src/main/res/values/strings.xml [hunk 15]
```diff
@@ -116,4 +116,10 @@
     >Number must be between <xliff:g id="min_value">%1$s</xliff:g> and <xliff:g
             id="max_value"
         >%2$s</xliff:g></string>
+    <string name="optional_text">\u0020(optional)</string>
```
[the remark this answers is not in the document]
Comment (santosh-pingle) [reply datacapture/src/main/res/values/strings.xml]: done.

File: datacapture/src/main/res/values/strings.xml (hunk 2 above)
Comment (santosh-pingle) [reply datacapture/src/main/res/values/strings.xml]: done.

File: datacapture/src/androidTest/java/com/google/android/fhir/datacapture/views/QuestionnaireItemDialogMultiSelectViewHolderFactoryEspressoTest.kt [hunk 16]
```diff
@@ -372,6 +372,75 @@ class QuestionnaireItemDialogMultiSelectViewHolderFactoryEspressoTest {
...
+          validationResult = NotValidated,
+          answersChangedCallback = { _, _, _, _ -> },
+          questionViewTextConfiguration = QuestionTextConfiguration(showAsterisk = true)
+        )
+      )
+
+      assertThat(viewHolder.itemView.findViewById<TextView>(R.id.question).text.toString())
+        .isEqualTo("Question? *")
+    }
+  }
```
Comment (jingtang10) [inline datacapture/src/androidTest/java/com/google/android/fhir/datacapture/views/QuestionnaireItemDialogMultiSelectViewHolderFactoryEspressoTest.kt:421]: move this to before the required test case, and also add a test case where asterisk is not enabled and check there's no asterisk appended to the question text.

File: datacapture/src/androidTest/java/com/google/android/fhir/datacapture/views/QuestionnaireItemDialogMultiSelectViewHolderFactoryEspressoTest.kt [hunk 17]
```diff
@@ -372,6 +372,75 @@ class QuestionnaireItemDialogMultiSelectViewHolderFactoryEspressoTest {
...
+
+  @Test
+  fun shows_optionalText() {
```
Comment (jingtang10) [inline datacapture/src/androidTest/java/com/google/android/fhir/datacapture/views/QuestionnaireItemDialogMultiSelectViewHolderFactoryEspressoTest.kt:424]: add a test case where optional / required are not enabled

File: datacapture/src/main/java/com/google/android/fhir/datacapture/extensions/MoreHeaderViews.kt [hunk 18]
```diff
@@ -68,3 +71,24 @@ internal fun initHelpViews(
   }
   helpTextView.updateTextAndVisibility(questionnaireItem.localizedHelpSpanned)
 }
+
+/**
+ * Appends ' *' to [Questionnaire.QuestionnaireItemComponent.localizedTextSpanned] text if
+ * [Questionnaire.QuestionnaireItemComponent.required] is true.
+ */
+internal fun appendAsteriskToQuestionText(
+  questionTextView: TextView,
```
Comment (jingtang10) [inline datacapture/src/main/java/com/google/android/fhir/datacapture/extensions/MoreHeaderViews.kt:80]: remove

File: datacapture/src/main/java/com/google/android/fhir/datacapture/views/QuestionnaireViewItem.kt [hunk 19]
```diff
@@ -81,7 +82,9 @@ data class QuestionnaireViewItem(
       emptyList()
     },
   internal val draftAnswer: Any? = null,
-  internal val enabledDisplayItems: List<Questionnaire.QuestionnaireItemComponent> = emptyList()
+  internal val enabledDisplayItems: List<Questionnaire.QuestionnaireItemComponent> = emptyList(),
+  internal val questionViewTextConfiguration: QuestionTextConfiguration =
+    QuestionTextConfiguration()
```
Comment (jingtang10) [inline datacapture/src/main/java/com/google/android/fhir/datacapture/views/QuestionnaireViewItem.kt:87]:
```suggestion
    QuestionTextConfiguration(),
```

File: datacapture/src/test/java/com/google/android/fhir/datacapture/views/factories/CheckBoxGroupViewHolderFactoryTest.kt [hunk 20]
```diff
@@ -470,4 +475,75 @@ class CheckBoxGroupViewHolderFactoryTest {
       )
       .isFalse()
   }
+
+  @Test
+  fun `show required text`() {
```
Comment (jingtang10) [inline datacapture/src/test/java/com/google/android/fhir/datacapture/views/factories/CheckBoxGroupViewHolderFactoryTest.kt:480]: add a case where there's no required or optional or asterisk.

and add that for every other test class here.

**Review by jingtang10 — APPROVED** [review]

File: datacapture/src/main/res/values/strings.xml [hunk 21]
```diff
@@ -116,4 +116,10 @@
...
+    <string name="optional_helper_text">Optional</string>
```
[the remark this answers is not in the document]
Comment (santosh-pingle) [reply datacapture/src/main/res/values/strings.xml:120]: At the moment, `Optional` text is not shown in the error message, therefore it does not require \n to separate out rest of the text in the error message.

File: datacapture/src/androidTest/java/com/google/android/fhir/datacapture/views/QuestionnaireItemDialogMultiSelectViewHolderFactoryEspressoTest.kt (hunk 16 above)
Comment (santosh-pingle) [reply datacapture/src/androidTest/java/com/google/android/fhir/datacapture/views/QuestionnaireItemDialogMultiSelectViewHolderFactoryEspressoTest.kt]: done.

File: datacapture/src/main/java/com/google/android/fhir/datacapture/views/QuestionnaireViewItem.kt (hunk 19 above)
Comment (santosh-pingle) [reply datacapture/src/main/java/com/google/android/fhir/datacapture/views/QuestionnaireViewItem.kt]: done.

File: datacapture/src/test/java/com/google/android/fhir/datacapture/views/factories/CheckBoxGroupViewHolderFactoryTest.kt (hunk 20 above)
Comment (santosh-pingle) [reply datacapture/src/test/java/com/google/android/fhir/datacapture/views/factories/CheckBoxGroupViewHolderFactoryTest.kt:519]: done.

**Pull request merged**